File: javascript/perfect-numbers/perfect-numbers.js
const isNaturalNumber = num => num > 0 && Number.isInteger(num);
const factors = num => {
  let factors = [];
  let divisor = num / 2;
  while (divisor > 0) {
    if (!(num % divisor)) factors.push(divisor);
    divisor--;
  }
  return factors;
}

export const classify = num => {
  if (isNaturalNumber(num)) {
    let aliquotSum = factors(num).reduce((sum, curr) => sum + curr, 0);
    if (aliquotSum === num) return 'perfect';
    if (aliquotSum < num) return 'deficient';
    if (aliquotSum > num) return 'abundant';
    return undefined;
  } else {
    throw new Error('Classification is only possible for natural numbers.');
  }
};

File: python/perfect-numbers/perfect_numbers.py
def classify(number):
  if number <= 0:
    raise ValueError('Not a natural number')
  factors = []
  if number > 1:
    factors.append(1)
  i = 2
  while i <= number / 2:
    if number % i == 0:
      factors.append(i)
    i += 1
  total = sum(factors)
  print(number, factors, sum(factors))
  if total == number:
    return 'perfect'
  elif total < number:
    return 'deficient'
  else:
    return 'abundant'
